graph.py

```diff
@@ -1,125 +1,116 @@
-import os
```

Comment:
This file was entirely rewritten using the previous one side by side to know what functionality had to be replicated, best to review as the whole file, not just a diff